core/container.go

```diff
@@ -332,7 +332,35 @@ func (container *Container) From(ctx context.Context, gw bkgw.Client, addr strin
 
 const defaultDockerfileName = "Dockerfile"
 
-func (container *Container) Build(ctx context.Context, gw bkgw.Client, context *Directory, dockerfile string, buildArgs []BuildArg, target string, secrets []SecretID) (*Container, error) {
+var buildCache = newCacheMap[uint64, *Container]()
+
+func cacheKey(keys ...any) uint64 {
+	hash := xxh3.New()
+
+	enc := json.NewEncoder(hash)
+	for _, key := range keys {
+		enc.Encode(key)
+	}
+
+	return hash.Sum64()
+}
+
+func (container *Container) Build(
+	ctx context.Context,
+	gw bkgw.Client,
+	context *Directory,
+	dockerfile string,
+	buildArgs []BuildArg,
+	target string,
+	secrets []SecretID,
+) (*Container, error) {
+	cached, initializer, found := buildCache.GetOrInitialize(cacheKey(container, context, dockerfile, buildArgs, target, secrets))
```

Comment:
You're right that two Build calls with different-ordered args, but same name/value pairs and all else being equal, are semantically equivalent. We could sort here, but I think it would be hard to handle that generically e.g. in `ToResolver` where the `--debug` vertex is created. To me it's not worth the trouble, since that seems like a super unlikely situation, and they'd at least be individually cached, and the order is at least stable.